### PR TITLE
Port to ARMv8 A64

### PR DESCRIPTION
**Pull request by hir12111**
This includes the `switch_frame` and `create_frame` functions for ARMv8 A64. It has been tested in a Raspberry Pi 4, but more tests would be nice.

### CODE REVIEW

File: context/switch-arm64.c [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+#endif
+
+__asm__(
+"       .text\n"
+
+// void * switch_frame(frame_t *old_frame, frame_t new_frame, void *arg)
+// Arguments on entry:
+//   x0      address of frame to be saved
+//   x1      frame to be loaded
+//   x2      Context argument to pass through
```
Comment (Araneidae) [inline context/switch-arm64.c:45]: Can you reference the relevant ABI documentation, please?  Have a look at `switch-x86.c` and `switch-x86_64.c` for examples of this.  I'd like to double check the registers you've chosen to save and restore

File: context/switch-arm64.c [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+ *      Chilton,
+ *      Didcot,
+ *      Oxfordshire,
+ *      OX11 0DE
+ *      michael.abbott@diamond.ac.uk
+ */
+
+// Coroutine frame switching for ARMv8 64 bits
+
+#ifdef __ARM_NEON
```
Comment (Araneidae) [inline context/switch-arm64.c:32]: Is NEON optional in the aarch64 ABI?

**Review by Araneidae — COMMENTED** [review]
Nice work.  I presume this is thoroughly tested?  I guess you have a Python build on aarm64?  How did you test it?

**Comment by Araneidae** [comment]
Oh, looks like CI failed.  Guess need to fix this first!  @thomascobb might have more understanding of what's failing here.

**Comment by thomascobb** [comment]
Ideally we would port this from travis to github actions, from a softioc binary to epicscorelibs.ioc, and build wheels as well as sdist for pypi.

This is what I'm doing to python IOC: https://github.com/dls-controls/pythonIoc/pull/5

Unfortunately I don't have time to do this for cothread...

**Comment by thomascobb** [comment]
Looks like you can emulate aarch64 with cibuildwheel:
https://cibuildwheel.readthedocs.io/en/stable/faq/#emulation

I'm using cibuildwheel in pythonIoc to build the wheel and run tests, but haven't done anything with emulation.

File: context/switch-arm64.c (hunk 1 above)
Comment (hir12111) [reply context/switch-arm64.c:69]: Done, I added a bit explanation and a link to the [Procedure Call Standard for the Arm 64-bit Architecture document](https://github.com/ARM-software/abi-aa/releases/download/2021Q1/aapcs64.pdf)

File: context/switch-arm64.c (hunk 2 above)
Comment (hir12111) [reply context/switch-arm64.c]: You are right, the ABI considers the associated registers but I thought some toolchain could decide to disable NEON, anyways, I've rebased to remove that part

**Comment by hir12111** [comment]
Regarding your question of how I tested it, it's not easy to explain, we can talk about it tomorrow

**Comment by hir12111** [comment]
> Looks like you can emulate aarch64 with cibuildwheel:
> https://cibuildwheel.readthedocs.io/en/stable/faq/#emulation

Very nice trick! using binfmt_misc to run binaries using qemu transparently, I suppose we could use that for running the unit tests

**Pull request merged**